modules/channel-teams/src/backend/client.ts

```diff
@@ -0,0 +1,181 @@
+import { BotFrameworkAdapter, ConversationReference, CardFactory, TurnContext, AttachmentLayoutTypes } from 'botbuilder'
+
+import * as sdk from 'botpress/sdk'
+import { Router } from 'express'
+import _ from 'lodash'
+
+import { Config } from '../config'
+
+import { Clients } from './typings'
+
+const outgoingTypes = ['message', 'typing', 'carousel', 'text']
+
+export class TeamsClient {
+  private conversationsRefs: _.Dictionary<Partial<ConversationReference>> = {}
+  private adapter: BotFrameworkAdapter
+  constructor(private bp: typeof sdk, private botId: string, private config: Config, private router: Router) {}
+
+  async initialize() {
+    this.adapter = new BotFrameworkAdapter({
+      appId: this.config.microsoftAppId,
+      appPassword: this.config.microsoftAppPassword
+    })
+
+    this.router.post('/api/messages', async (req, res) => {
+      await this.adapter.processActivity(req, res, async turnContext => {
+        const conversationReference = TurnContext.getConversationReference(turnContext.activity)
+        if (!turnContext.activity.text) {
+          // To prevent from emojis reactions to launch actual events
+          return
+        }
+
+        const threadId = conversationReference.conversation.id
+        this.conversationsRefs[threadId] = conversationReference
+        this._sendIncomingEvent(turnContext, threadId, { type: turnContext.activity.type })
```

Comment:
Not sure to understand this conversationsRef part, what is it used for ? If the bot runs on more than one node, what happens if a message gets dispatched to node 1, and others to node 2 ?

Reply:
**Not sure to understand this conversationsRef part, what is it used for ?**

The problem is that the botframework requires to answer each message inside the adapter.processActivity function. If not, the connection gets unusable (Can't remember the exact error message). 

The only work around I found was to save a reference of the conversation inside a data structure to be used latter.

This allows to send an event inside the whole middleware/hooks chain and only when our outgoing handler is called, to send back the outgoing event using the conversation reference.


**what happens if a message gets dispatched to node 1, and others to node 2 ?**

I'm not 100% sure as I have not tested it, but I believe it will still work. This mecanism only ensure an outgoing event will be sent back using the same connection that it arrived from.

It basically works like this:

```
let lastConnection = null

router.post('api/messages', connection => {
    lastConnection = connection
})

function outgoingMiddlewareHandler(event) {
    lastConnection.sendBack(event)
}
```

Reply:
https://docs.microsoft.com/en-us/javascript/api/botbuilder/botframeworkadapter?view=botbuilder-ts-latest#continueconversation-partial-conversationreference----context--turncontext-----promise-void--

Reply:
This documentation helped me out:

https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-howto-proactive-message?view=azure-bot-service-4.0&tabs=javascript